File: data-structures/01_Arrays/vector.py
import array as arr


class Vector:

    def __init__(self):
        self.__size = 0
        self.__capacity = 1
        self.__elements = arr.array('i', [0] * self.__capacity)

    # Check the size of the vector
    def size(self):
        return self.__size

    # Return the capacity of the vector
    def capacity(self):
        return self.__capacity

    # Check if the array is empty or not
    def is_empty(self):
        return (self.__size == 0)

    # Return an element at index
    def at(self, index):
        return self.__elements[index]

    def print_array(self):
        print(self.__elements)
        print("Size: ", self.__size, " Capacity: ", self.__capacity)

    def __resize(self, new_capacity):

        # print("Out of space. Resizing the array.")
        new_array = arr.array('i', [0] * new_capacity)

        # Increase the capacity
        self.__capacity = new_capacity

        # Copy the old array to the new array
        for i in range(self.__size):
            new_array[i] = self.__elements[i]

        return new_array

    def push(self, item):
        # print("Pushing ", item)
        index = self.__size
        # Check if there is space in the array
        if (self.__size < self.__capacity - 1):
            # print("Inserting at index:", index )
            self.__elements[index] = item
            self.__size += 1
        else:
            # Increase the capacity of the array
            self.__elements = self.__resize(self.__capacity * 2)

            # print("Inserting at index:", index)

            self.__elements[index] = item
            self.__size += 1

        self.print_array()

    def insert(self, index, item):
        print("Inserting ", item, " at index: ", index)

        if (index > self.__size):
            print("Array out of bounds. Unable to insert.")
        # See if we are inserting at the tail end.
        elif (index == self.__capacity - 1):
            self.push(item)
        else:
            # Check if we have space to insert the new item
            if (self.__size + 1 >= self.__capacity):
                self.__elements = self.__resize(self.__capacity * 2)

                # shift all elements 1 space to the right
                for i in range(self.__size + 1, index, -1):
                    self.__elements[i] = self.__elements[i-1]

                self.__elements[index] = item
                self.__size += 1

            self.print_array()

    def prepend(self, item):
        # This is same as inserting at index 0
        self.insert(0, item)

    def pop(self):
        index = self.__size - 1
        popped_value = self.__elements[index]
        self.__elements[index] = 0
        self.__size -= 1

        # See if we need to resize the array
        if (self.__size <= self.__capacity / 4):
            self.__elements = self.__resize(int(self.__capacity / 2))

        self.print_array()

        return popped_value

    def find(self, item):

        for i in range(self.__size):
            if(self.__elements[i] == item):
                return i

        return -1

    def delete(self, index):

        if (index >= self.__size):
            print("Array out of bounds. Unable to delete element at index: ",
                  index)
        else:

            # shift all the elements to the left by one
            for i in range(index, self.__size):
                self.__elements[i] = self.__elements[i+1]

            print("Deleting element at index: ", index)
            self.__elements[self.__size] = 0
            self.__size -= 1

            # See if we need to resize the array
            if (self.__size <= self.__capacity / 4):
                self.__elements = self.__resize(int(self.__capacity / 2))

            self.print_array()

    def remove(self, item):

        # Look for all instances of item in the array and remove
        more_items = True

        print("Deleting all instances of ", item)

        while(more_items):
            item_index = self.find(item)
            if (item_index >= 0):
                self.delete(item_index)
            else:
                more_items = False


# test code
v = Vector()
v.push(3)
v.prepend(2)

File: data-structures/01_Arrays/Vector.cs
class Vector
{
	private int[] elements;
	private int size;
	private int capacity;
	
	private Vector()
	{
		elements = new int[1];
		size = 0;
		capacity = 1;
	}
	
	public int Size()
	{
		return size;
	}
		
	public int Capacity()
	{
		return capacity;
	}
		
	public bool IsEmpty()
	{
		return (size == 0);
	}
	
	public int At(int index)
	{
		return elements[index];
	}
	
	private int[] Resize(int new_capacity)
	{
		int[] new_elements = new int[new_capacity];
				
		// copy the old array to the new array
		for(int i = 0; i < size; i++)
			new_elements[i] = elements[i];
		
		capacity = new_capacity;
		
		return new_elements;
	}
	
	// insert an item at the end of the vector
	public void Push(int item)
	{
		if(size == capacity)
			elements = Resize(capacity * 2);
		
		elements[size] = item;
		size++;
	}

	public void Insert(int index, int item)
	{
		if (size + 1 == capacity)
			elements = Resize(capacity * 2);
		
		// Shift all the elements after the index to the right to free up space
		for(int i = size; i > index; i--)
			elements[i] = elements[i-1];
		
		// insert the new item at index
		elements[index] = item;
		size++;
	}
	
	public void Prepend(int item)
	{
		Insert(0, item);
	}
	
	
	public void Display()
	{
		foreach(int element in elements)
				System.Console.Write("{0} ", element);
				
		System.Console.WriteLine();
	}
	
	public int Pop()
	{
		int index = size - 1;		
		int item = elements[index];
		
		elements[index] = 0;
		size--;
		
		// Resize the array if necessary
		if (size == capacity / 4)
			elements = Resize(capacity / 2);
		
		return item;
	}
	
	public void Delete(int index)
	{
		// Shift all elements to the left
		for(int i = index; i < size; i++)
			elements[i] = elements[i+1];
		
		size--;
		elements[size] = 0;
		
		// Resize the array if necessary
		if (size == capacity / 4)
			elements = Resize(capacity / 2);
	}
	
	public int Find(int item)
	{
		for(int i=0; i < size; i++)
				if(elements[i] == item)
					return i;
				
		return -1;		
	}
	
	public void Remove(int item)
	{
		int index = Find(item);		
		
		while(index > -1)
		{
			Delete(index);
			index = Find(item);
		}
	}
	
	
	
	public static void Main(string[] args)	
	{
		Vector v = new Vector();
		
		System.Console.WriteLine("Vector size: {0}", v.Size());
		System.Console.WriteLine("Vector capacity: {0}", v.Capacity());
		System.Console.WriteLine("Is the vector empty: {0}", v.IsEmpty());
		
		v.Push(1);
		v.Display();
		
		v.Push(99);
		v.Display();
		
		v.Push(99);
		v.Display();
		
		v.Push(99);
		v.Display();
		
		v.Push(5);
		v.Display();
		
		v.Insert(0, 99);
		v.Display();
		
		v.Insert(2, 100);
		v.Display();
		
		v.Insert(5, 99);
		v.Display();
		
		v.Prepend(99);
		v.Display();
		
		System.Console.WriteLine("Element {0} found at index: {1}", 100, v.Find(100));
		v.Display();
		
		v.Remove(99);
		v.Display();
		
		System.Console.WriteLine("Vector size: {0}", v.Size());
		System.Console.WriteLine("Vector capacity: {0}", v.Capacity());
		System.Console.WriteLine("Is the vector empty: {0}", v.IsEmpty());
		
		
	}
	
	
}